Remove the CAN_INSTALL file when occ maintenance:install is complete

alexharpin opened the pull request:
When occ maintenance:install is run from the CLI, the CAN_INSTALL in the config directory is left in place when installed is set to true, whereas this file is removed when the web installer is used.  Removing this file in the CLI command maintains consistency between the two.  This allows automation tools an easier way to determine if this process has been completed.

Signed-off-by: Alex Harpin <development@landsofshadow.co.uk>

alexharpin committed:
Remove the CAN_INSTALL file when occ maintenance:install is complete

When occ maintenance:install is run from the CLI, the CAN_INSTALL in the config directory is left in place when installed is set to true, whereas this file is removed when the web installer is used.  Removing this file in the CLI command maintains consistency between the two.  This allows automation tools an easier way to determine if this process has been completed.

Signed-off-by: Alex Harpin <development@landsofshadow.co.uk>

cyclops8456 commented:
I came across this when creating a Kubernetes deployment for Nextcloud, with a batch job to perform the initial setup.  I'd expected the CAN_INSTALL to be removed when the setup had completed, in the same manner as the web installer, so was using the presence of this file in an init container to delay the deployment of the frontend Apache container.  Worked around this issue by delete the file in the batch job instead for now.

cyclops8456 commented:
Resolves #24301

cyclops8456 commented:
Looks like a few of the CI tests failed due to git clone issues :(

juliushaertl commented:
Good catch, would be good to also check the return value of unlink and log an error in case the file was not removed.

cyclops8456 commented:
@juliushaertl Would you class this enough of an error to return it in the error array from the install method in the Setup class, which would then be handled in the install command, returning with an exit code of 1 instead of 0 (core/Command/Maintenance/Install.php:109), or just log the fact it couldn't be removed?

alexharpin committed:
Move CAN_INSTALL check to method and remove unlink from SetupController

Move the check for the CAN_INSTALL file in the config directory to a method in the Setup class and remove the call to unlink from the SetupController as this in now handled in the Setup class.

Signed-off-by: Alex Harpin <development@landsofshadow.co.uk>

alexharpin committed:
Log a warning to CLI install command if CAN_INSTALL is not removed

Log a warning for the CLI install command if the CAN_INSTALL file still exists at the end of the installation.  This matches the warning logged by the web installer.

Signed-off-by: Alex Harpin <development@landsofshadow.co.uk>

alexharpin committed:
Rename canInstallExists method and add new method for removal

Rename canInstallExists to shouldRemoveCanInstallFile to cover removal of this file for non-git channels and logging any failure to remove it.

Add new method to detect if this file exists during web based installation.

Signed-off-by: Alex Harpin <development@landsofshadow.co.uk>

cyclops8456 commented:
Little bit of a rework as the SetupController was calling the install method of the Setup class, so was duplicating the check to unlink the CAN_INSTALL file.  Resolved that and added logging to the CLI command to output a warning that the file needs to be removed manually (matching the web installer).

juliushaertl commented on core/Controller/SetupController.php:
```diff
@@ -112,10 +112,8 @@ private function finishSetup(bool $installRecommended) {
-			if (!unlink(\OC::$configDir.'/CAN_INSTALL')) {
-				\OC_Template::printGuestPage('', 'installation_incomplete');
-			}
+		if ($this->setupHelper->shouldRemoveCanInstallFile()) {
```
The unlink is no longer triggered now, I'd probably add another method removeCanInstallFile() for the unlink which returns a boolean then

cyclops8456 replied:
@juliushaertl The unlinking is now handled in the setupHelper, rather than the same operation being performed in multiple places.

juliushaertl commented on core/Command/Maintenance/Install.php:
```diff
@@ -110,6 +110,9 @@ protected function execute(InputInterface $input, OutputInterface $output): int
 			$this->printErrors($output, $errors);
 			return 1;
 		}
+		if ($setupHelper->shouldRemoveCanInstallFile()) {
```
Might be good to also make sure that the install is blocked at the beginning of the command similar to https://github.com/nextcloud/server/pull/27492/files#diff-5f51d6576e5fd6272d9c29288da65fee6e760b25f016bccdef84479a4bbaf0b8R67

cyclops8456 replied:
@juliushaertl I didn't include this initially as the command is only registered when the system installed value is set to false, so the command isn't available to run at all; so the check seems redundant at this point.  Does this still want including in this case?

szaimen commented:
@cyclops8456 any update here?

szaimen commented:
@cyclops8456 any update here?

szaimen commented:
Would this be a solution to https://github.com/nextcloud/server/issues/18620 ?

cyclops8456 commented:
@szaimen Regarding #18620, this wouldn't have resolved that particular case, but the fix for that could still be impacted by the non-removal of the CAN_INSTALL file, at least for CLI installation (so more for automated installs).

cyclops8456 commented:
@szaimen @juliushaertl Returning to this MR after getting snowed under with other work and completely forgetting about it.  I've rebased my branch against current master and have tested it locally for both web and CLI based installs and the CAN_INSTALL file is removed as expected in both cases now.  This should mean that automated installs don't need to do this manually now, and I can remove it from my Kubernetes deployment config.

szaimen commented:
failing CI likely unrelated
